### ISSUE TITLE
13sdawad

### ISSUE THREAD

**Issue by davalillo-rm** [davalillo-rm/davalillo-rm#149]
import pandas as pd

# 1. Leemos tu fuente oficial
df_raw = pd.read_excel('Data_historica.xlsx')

# 2. Creamos los "Buckets" (agrupamos por plazos parecidos)
# Esto es necesario porque no siempre hay un bono de exactamente 30 días
def asignar_bucket(dias):
    if 25 <= dias <= 35: return '1M'
    if 85 <= dias <= 95: return '3M'
    if 175 <= dias <= 185: return '6M'
    if 355 <= dias <= 365: return '1Y'
    return None

df_raw['Bucket'] = df_raw['Dias'].apply(asignar_bucket)

# 3. Transformamos a formato WIDE (Lo que el PCA necesita)
# Filtramos solo los que entraron en un bucket y pivotamos
df_pca_input = df_raw.dropna(subset=['Bucket']).pivot_table(
    index='Fecha', 
    columns='Bucket', 
    values='Yield', 
    aggfunc='mean'
).dropna() # Eliminamos días donde falte algún nodo para tener la curva completa

print("¡Matriz preparada para el PCA!")
print(df_pca_input.head())

**Comment by davalillo-rm** [comment]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA

# 1. CARGAMOS TU MATRIZ PREPARADA (La que creamos en el paso anterior)
# Asumimos que df_pca_input ya tiene las columnas ['1M', '3M', '6M', '1Y']
# Y que las filas son las Fechas.

# IMPORTANTE: Trabajamos con los cambios diarios (Diferencias)
# El PCA sobre niveles absolutos suele fallar por falta de estacionariedad.
df_diff = df_pca_input.diff().dropna()

# 2. ENTRENAMIENTO DEL MODELO
# Extraemos los 3 componentes principales
pca = PCA(n_components=3)
pca.fit(df_diff)

# 3. ANALIZAMOS EL PODER EXPLICATIVO
varianza_explicada = pca.explained_variance_ratio_ * 100
print(f"Poder del PC1 (Nivel): {varianza_explicada[0]:.2f}%")
print(f"Poder del PC2 (Pendiente): {varianza_explicada[1]:.2f}%")
print(f"Poder del PC3 (Curvatura): {varianza_explicada[2]:.2f}%")

# 4. GRAFICAMOS LOS 'LOADINGS' (Los vectores propios)
# Esto nos dice cómo reacciona cada nodo a cada fuerza.
nodos = df_pca_input.columns
loadings = pca.components_



plt.figure(figsize=(10, 6))
plt.plot(nodos, loadings[0], marker='o', label=f'PC1 (Nivel) - {varianza_explicada[0]:.1f}%')
plt.plot(nodos, loadings[1], marker='s', label=f'PC2 (Pendiente) - {varianza_explicada[1]:.1f}%')
plt.plot(nodos, loadings[2], marker='^', label=f'PC3 (Curvatura) - {varianza_explicada[2]:.1f}%')

plt.axhline(0, color='black', linestyle='--', alpha=0.5)
plt.title('Interpretación de los Componentes Principales (Curva CDBCRP)')
plt.xlabel('Plazo (Nodos)')
plt.ylabel('Sensibilidad / Peso')
plt.legend()
plt.grid(alpha=0.3)
plt.show()
